Add SpriteUtil to API package

IThundxr opened the pull request.

IThundxr committed:
Convert spriteUtil into a api

jellysquid3 commented on common/src/api/java/net/caffeinemc/mods/sodium/api/texture/SpriteUtil.java:
```diff
@@ -0,0 +1,26 @@
+     * Marks a provided sprite as active
+     * @param sprite The sprite you want to mark as active
+     */
+    void markSpriteActive(@Nullable TextureAtlasSprite sprite);
```
I'm not really convinced there's any reason a mod should be calling this function with `null`. We should make the parameter `@NotNull` before promoting it to API status.

IThundxr replied:
Is this not a issue anymore?
> Can happen in some cases, for example if a mod passes a BakedQuad with a null sprite to a VertexConsumer that does not have a texture element.

jellysquid3 replied:
It may be an issue. But in my opinion it should be on the caller to handle. Having an API where the method contract is "this will do nothing if null is passed" is strange.

jellysquid3 commented on common/src/api/java/net/caffeinemc/mods/sodium/api/texture/SpriteUtil.java:
```diff
@@ -0,0 +1,26 @@
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+@ApiStatus.AvailableSince("0.6.0")
```
Needs incubating status.

IThundxr committed:
Make it experimental

IThundxr committed:
Require NotNull

jellysquid3 commented on common/src/api/java/net/caffeinemc/mods/sodium/api/texture/SpriteUtil.java:
```diff
@@ -0,0 +1,27 @@
+     * @param sprite The sprite you want to check
+     * @return true if the provided sprite has an animation, false if it doesn't
+     */
+    boolean hasAnimation(TextureAtlasSprite sprite);
```
Both functions need to have their parameters annotated with `@NotNull.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/client/render/frapi/render/NonTerrainBlockRenderContext.java:
```diff
@@ -127,6 +127,6 @@ protected void shadeQuad(MutableQuadViewImpl quad, LightMode lightMode, boolean
     private void bufferQuad(MutableQuadViewImpl quad) {
         QuadEncoder.writeQuadVertices(quad, vertexConsumer, overlay, matPosition, trustedNormals, matNormal);
-        SpriteUtil.markSpriteActive(quad.sprite(SpriteFinderCache.forBlockAtlas()));
+        SpriteUtil.INSTANCE.markSpriteActive(quad.sprite(SpriteFinderCache.forBlockAtlas()));
```
This can return null.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/immediate/buffer_builder/intrinsics/BufferBuilderMixin.java:
```diff
@@ -24,8 +24,8 @@ public abstract class BufferBuilderMixin implements VertexConsumer {
-
-            SpriteUtil.markSpriteActive(bakedQuad.getSprite());
+            
+            SpriteUtil.INSTANCE.markSpriteActive(bakedQuad.getSprite());
```
Quads are not guaranteed to have sprites, so this can pass null to the function.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/immediate/buffer_builder/intrinsics/BufferBuilderMixin.java:
```diff
@@ -41,15 +41,16 @@ public void putBulkData(PoseStack.Pose matrices, BakedQuad bakedQuad, float r, f
         BakedModelEncoder.writeQuadVertices(writer, matrices, quad, color, light, overlay);
 
-        SpriteUtil.markSpriteActive(quad.getSprite());
+        if (quad.getSprite() != null)
```
Code style nitpick: All if-statements need brackets.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/immediate/buffer_builder/intrinsics/BufferBuilderMixin.java:
```diff
@@ -41,15 +41,16 @@ public void putBulkData(PoseStack.Pose matrices, BakedQuad bakedQuad, float r, f
 
-        SpriteUtil.markSpriteActive(quad.getSprite());
+        if (quad.getSprite() != null)
+            SpriteUtil.INSTANCE.markSpriteActive(quad.getSprite());
```
`quad.getSprite()` should be pulled into a local variable so that it cannot change between the null check and passing it to the function.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/immediate/buffer_builder/intrinsics/BufferBuilderMixin.java:
```diff
@@ -41,15 +41,16 @@ public void putBulkData(PoseStack.Pose matrices, BakedQuad bakedQuad, float r, f
-
-            SpriteUtil.markSpriteActive(bakedQuad.getSprite());
+            
+            SpriteUtil.INSTANCE.markSpriteActive(bakedQuad.getSprite());
```
Also needs a null check.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/immediate/buffer_builder/intrinsics/BufferBuilderMixin.java:
```diff
@@ -64,6 +65,7 @@ public void putBulkData(PoseStack.Pose matrices, BakedQuad bakedQuad, float[] br
         BakedModelEncoder.writeQuadVertices(writer, matrices, quad, r, g, b, a, brightnessTable, colorize, light, overlay);
 
-        SpriteUtil.markSpriteActive(quad.getSprite());
+        if (quad.getSprite() != null)
```
See prior comments about memory safety & if-statements.

jellysquid3 commented on common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/model/item/ItemRendererMixin.java:
```diff
@@ -94,7 +94,8 @@ private void renderBakedItemQuads(PoseStack.Pose matrices, VertexBufferWriter wr
             BakedModelEncoder.writeQuadVertices(writer, matrices, quad, color, light, overlay);
 
-            SpriteUtil.markSpriteActive(quad.getSprite());
+            if (quad.getSprite() != null)
```
See prior comments about null safety.

jellysquid3 commented on fabric/src/main/java/net/caffeinemc/mods/sodium/mixin/fabric/features/render/model/block/ModelBlockRendererMixin.java:
```diff
@@ -46,7 +46,8 @@ private static void renderQuads(PoseStack.Pose matrices, VertexBufferWriter writ
             BakedModelEncoder.writeQuadVertices(writer, matrices, quad, color, light, overlay);
 
-            SpriteUtil.markSpriteActive(quad.getSprite());
+            if (quad.getSprite() != null)
```
Code style nitpick... needs braces.

jellysquid3 commented on neoforge/src/main/java/net/caffeinemc/mods/sodium/neoforge/mixin/features/render/model/block/ModelBlockRendererMixin.java:
```diff
@@ -48,7 +48,8 @@ private static void renderQuads(PoseStack.Pose matrices, VertexBufferWriter writ
             BakedModelEncoder.writeQuadVertices(writer, matrices, quad, color, light, overlay);
 
-            SpriteUtil.markSpriteActive(quad.getSprite());
+            if (quad.getSprite() != null)
```
Ditto.

IThundxr committed:
address reviews

IThundxr committed:
Fix imports

IThundxr committed:
Merge remote-tracking branch 'upstream/1.21.1/stable' into spriteutil

IThundxr committed:
Merge branch '1.21.1/stable' into spriteutil

jellysquid3 committed:
Adjust texcoord epsilon depending on atlas size

This fixes some texture precision issues with very large
atlases, and with hardware that has limited sub-texel
precision.

jellysquid3 committed:
Bump version to 0.6.7-rc.1

jellysquid3 committed:
Use 8 bits of sub-texel precision and fix off-by-one error

IThundxr committed:
Merge remote-tracking branch 'upstream/1.21.1/stable' into spriteutil

# Conflicts:
#	common/src/main/java/net/caffeinemc/mods/sodium/mixin/features/render/model/item/ItemRendererMixin.java

IThundxr committed:
Merge branch '1.21.1/stable' into spriteutil
